### ISSUE TITLE
Make action types and reactors externally composible

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#109]
See https://github.com/WordPoints/hooks-api/issues/108#issuecomment-194310828:

> One side-effect of using the same names for action types from different event[ type]s was that it allowed us to reference a single response type from the reactors across different types of events. The toggle events and repeatable events both wanted the same response from the points reactor, for example, so they both called that action type by a single name.
> 
> With split action types however, we can no longer do that. This becomes an issue if and when we introduce a new event type, because it will have action types that no previously-created reactors would be able to listen for. Before we could just call its action type the same thing as one of the existing action types, and reactors would automatically give that same response. But now we can't do that.
> 
> So it seems that we need to make the reactors and action types more composible. Instead of the reactor composing them, perhaps there should be a registry or something (maybe on the router like with the actions?) and they should be hooked up externally. That would allow for the reactors to be configured for types of events that they didn't know about, and also give full flexibility in configuring how those reactors should respond. So we could match up the fire event with the reverse response, and vice versa, if we wanted to.
> 
> That is assuming, of course, that we can also compose the reactor responses, which isn't currently the case. So I guess that would also have to be composed externally, rather than internal to each reactor as it currently is. As things stand, each reactor has to check for different action types internally and route them to different responses if that's what it wants to do. But I guess it makes more sense for that to be in the router, doesn't it?
> 
> Even now that we're thinking of moving the composing of responses out of the reactors though, we're still just assuming that this should be a per-reactor thing. But maybe there is a case to be made that each reaction should be allowed to control this if it wants to. After all, reactions already have to know about the event they are attached to, so they have to know about the action types that accompany it (or at least those that they want to react to). So it makes sense that reactions should be able to tell the router how to route action types to reactor responses.
> 
> I've actually been thinking about how reactions listen to events, and not action types (or firers, previously), but that perhaps that should change. And I think it is starting to make a lot of sense. I think it would also make sense to revisit some things that we discussed #104, like splitting the reactors so that the each have just one response, though there is still a lot of things to work out there.

**Comment by JDGrimes** [comment]
## Single-response reactors

Composing reactors and their responses was discussed at length in https://github.com/WordPoints/hooks-api/issues/104#issuecomment-187940983.

One thing that was discussed there was whether reactors should really be composing multiple responses, or whether each should define a single response, and these should be composed elsewhere (by the reactions, specifically). That is really for another ticket, but it fits nicely with the idea being explored in this ticket. One implication of that is that we'd then have to move the reaction stores away from the reactors, and make them reactor-agnostic instead of being closely tied to a reactor as they are now. I was thinking just now that the logic for handling the reaction stores could then be moved into the router, which I think makes a lot of sense. (The reaction store registry would still be its own thing, of course.)

**Comment by JDGrimes** [comment]
## Hit blocker extension

One interesting consequence of this change, if we allow the reactions themselves to do the composing, is that it would make the blocker extension just introduced in #102 unnecessary, because any reaction that didn't want to listen to a particular action type could just not define a response for that action type.

**Comment by JDGrimes** [comment]
## How would reactions compose this in the UI?

I've begun the work to extract this, but currently the composition happens at the router, which response/hit type a reactor should make for each type of action. I still think that letting the reaction choose the hit type sounds reasonable though, since we may want to ignore some hit types for some reactions anyway. But I'm still wondering how exactly it would work. Specifically, how would the reaction know within the UI which hit types it can choose from and how to pair them with action types?

I guess too that maybe there is a different between the reaction choosing to ignore a hit type and a the reaction choosing which hit types should be hooked to which action types. It would be possible, I suppose, to have the hit type/action type relationships composed elsewhere, but still leave it up to the reaction which hit types it references.

### Hit type or action type?

I say we'd be ignoring a _hit type_, but I'm actually not so sure about that. Do we really want to ignore a hit type or an action type?

If it really is a hit type that we want to ignore, we couldn't do that with the code as it currently stands. That is, we couldn't use an extension to do it as previously proposed, because the extensions aren't currently told what type of hit the hit will be.

I guess it used to be based on fire type, which is what became the action types. But then it was really about the hit type; for the points hooks, it was about ignoring all _reverses_, because we didn't want points removed at all. So for that it was indeed about hit types.

Anyway, that just means that if we want to do this with an extension, we at least need to modify the current code so that the hit type is specified on the fire object. (And there are sure to be other uses for that anyway.)

### How would hit types and action types be paired?

But the real question is still how the a hit type would be chosen within the UI, were we to let each reaction do the composing (and not just choosing which action type/hit type pair to listen to). I suppose that it would know what hit types the reactor supported, and what action types the event offered. (Though at the moment we'd have to modify things in order for it to be able to have that information, as the events don't offer a list of action types anywhere, and though the reactors do offer a list of hit types, they aren't made publicly available, and don't include possible custom hit types not represented as methods on the reactor object.) But how would it know which hit types to pair with which action types to achieve the desired result?

### Hard-coding per implementation

If it has to compose them entirely, it seems we'd end up having to hard-code that in each implementation. Otherwise the user would actually be doing the composing, and I don't think that's what we want (at least not for the points reaction stores). So ultimately this would be composed by the developer (or _perhaps_ within the UI). But that doesn't mean that it still couldn't be done through each reactions settings, though that might not be our only option. 

So we'd likely want to consider at least making it so that this is composed per-reaction store (or implementation, really), rather than per reactor, even if we still do that on the router instead of the reaction level. I think that really makes sense, since reactors can in theory be used for multiple different implementations, and might want to be used differently. Assuming of course that reactors really would ever be reused across different implementations and hooked to different action types. That's definitely more composibility than is currently necessary in the API, but I can see where it could possibly become useful in the future (think for example of a reactor that sends notifications of some sort, it could possibly have many different uses).

I'm not entirely convinced that we've actually answered the question satisfactorily as to how the action and hit types would be paired in the UI though, even assuming that this was hard-coded. Actually, how could it even be hard-coded, when new event types and action types could be added in the future? This does have to be composed somewhat dynamically then, unless we say that new event types would be such a rare thing that they could be added with an update of the hard-code itself. The point is, that it all really comes down to being per-implementation, whether the router determines this from a registry or a reaction setting or something else.

### Non hard-coding, per reaction

The benefit of letting the reaction settings ultimately determine this though, is that it allows greater flexibility. It doesn't force each implementation to hard-code this to work only a single way or else use workarounds in custom extensions to achieve alternate behavior. It makes it possible for a particular implementation to let the user make this choice in the UI if they want to.

We probably don't want to do that with the points hooks, but does that mean it shouldn't be possible for any other implementation to do that either if it wants to? (Actually, I guess we might say that we do want to do this in the points reactor, because we want to allow particular reactions to be arbitrarily unhooked from certain hit types for legacy reasons. Although as we said above, that is a separate issue from complete composition of action type to hit type.) In that case, the per-implementation pairing is something that should be implementation specific then, not something officially sanctioned by being part of the router. It might not be hard-coded exactly, but it should be maintained by each implementation as needed, rather than for all of them on the router, since the router wouldn't really be using this info anymore.

(And note that if an implementation supports multiple reactors, this would have to be per reactor per implementation, not just per implementation as we originally proposed. But if each implementation decides for itself how to do this, it can do it more simply if it only uses one reactor. Though of course if we did just do this for all implementations on the router we would need to make it per-reactor. And note that this is true even with a single reactor per reaction, since reaction stores aren't specifically tied to reactors anymore and can contain reactions for multiple different reactors.)

Actually, I'm not so sure that we don't need to reconsider how we do this in the UI for the points hooks themselves, because the state of the events API continues to evolve. Related is #110, #111, and #95. Are events supposed to work in both directions now? It think we need to redefine the nature of events, but that is something for another ticket.

**Comment by JDGrimes** [comment]
OK. After a bit of a hiatus here, we're back. The main reason that I'm looking at this right now is that there are some changes relating to this ticket still sitting in the router code. The commit c5f0016 was made, and then I left the [above comment](https://github.com/WordPoints/hooks-api/issues/109#issuecomment-197928937). The code that I have reflects the consideration in the above comment of making this per implementation/reaction store. The reason that it was never committed was likely because we had second thoughts about this approach at the end of that comment. We were considering the value of moving entirely to a per-reaction approach instead. We then went off on the tangent of #116, which at first was an attempt to decide whether events are now supposed to work both ways, but eventually led us to a full reconsideration of exactly what we're doing here and what direction the API is going. As a result of that discussion, it is likely that the future direction of the API will be very different than what we were originally envisioning here. The events API will one day only handle what we called "repeatable" events, whereas an entity change API will likely be introduced to handle the stateful events. However, I'm here for two reasons: 1) we need to decide how that transition is going to be made, and how much of the multi-purpose events API needs to be in the initial version, (#129) and 2) I need to clean up the router code and remove any uncommitted cruft so that we can begin to do that. So right now my most immediate concern is figuring out whether I should polish up and commit the per-implementation code currently sitting in the router or not.

**Comment by JDGrimes** [comment]
I've decided to go ahead and clean this up and commit it, if for nothing else then for a historical record of where we were at this point.

Looking over my unpublished notes on this, I see that there was something in regard to this that was yet to be discussed, and which I've sort of pondered over as I was looking back over this.

## Reaction Stores =? Implementations

The fact is that we have been essentially equating "implementations" and reaction store slugs. We've acknowledged the difference between them to some degree, but we haven't explained exactly why that difference isn't significant. That is, we haven't explained why it is OK for us to treat reaction store slugs as if they were implementation identifiers.

The reason that we're equating these two things is because "implementations" currently aren't a thing in the API, and reaction stores are as close as we come to them. Basically, the reaction store slugs _are_ implementation identifiers by another name. Each reaction store groups reactions that we want to display in a particular UI. This is just the easiest way to group the reactions per-implementation.

So the reaction store slug is something that we can already use in the router and relate to particular reactions. We even already log it in the hit logs—which is something that we have to do with implementation IDs now that they help define the particular response from a reactor for a particular hit.

**Comment by JDGrimes** [comment]
## Logging the hit type?

On the subject of the hit logs, another thing that I had wondered when I was working on this previously was whether we should be logging the hit types as well. At present though, there is a one-to-one correspondence between hit types and action types per reactor per reaction store. So logging the hit type in addition to the action type doesn't seem necessary.

On further reflection though, it does seem like logging the hit type would make querying the logs simpler, as well as being more future-proof and robust (assuming that we might ever want to change this one-to-one correspondence in the future). Why should we have to query by three parameters (action type, reactor slug, and reaction store slug), instead of just one? I suppose that we'd generally be querying via reactor slug anyway though, so it would only be one extra parameter then (the reaction store slug). Still, the most egregious part of this is that it would be very difficult to get the information needed to know which reaction store and action type we would want to query by, because the router's reactor hit type index will likely be indexed by action type and storage slug and then by reactor slug—with the hit type out at the very tip end.

So I guess if we want to ever query by hit type, we probably ought to log it.

But perhaps that is a fairly large "if". Is there ever a reason that we'd want to query by the hit type instead of the action type? That's a question that might take a bit of work to answer, and I'm loathe to do that when we might be getting ready to rip out hit types anyway.

**Comment by JDGrimes** [comment]
Actually, hit types don't have a one-to-one correspondence to action types. Each action type can currently only correspond to a single hit type (per reactor per reaction store), but multiple action types can trigger the same hit type.

Most obviously, this means that we'd sometimes have to query by multiple action types to include all logs for a single hit type.

But it is actually a bigger issue than that, because which action types we affect which hit types can then change not just when an action type changes to specify a different hit type (which is more of an edge cases), but also when new action types are registered, or existing action types are removed. That means that we can't be sure that the list of action types that trigger a particular hit type now is complete—other action types may have previously existed that also triggered that hit type.

All of this is still moot of course, if we don't need to query by hit type in the first place.

**Comment by JDGrimes** [comment]
Here is the code from the router. It includes two different possible approaches. One is only slightly modified from the current code, but the other (the first one) is completely rewritten. I've decided not to commit it right now, because it would end up only half-baked, and I don't want to take the time to flesh it all the way out (logging the hit type, etc.) since we're likely taking the API in an entirely different direction.

``` php

    /**
     * Fire an event at each of the reactions.
     *
     * @since 1.0.0
     *
     * @param string                     $action_type The type of action triggering
     *                                                this fire of this event.
     * @param string                     $event_slug  The slug of the event.
     * @param WordPoints_Hook_Event_Args $event_args  The event args.
     */
    public function fire_event(
        $action_type,
        $event_slug,
        WordPoints_Hook_Event_Args $event_args
    ) {

        // Needed as long as this method is public, at least.
        if ( empty( $this->reactor_index[ $action_type ] ) ) {
            return;
        }

        $hooks = wordpoints_hooks();

        // We thought that it was needless to pull up all of the reaction stores when
        // we only need those that have a hit type registered for this action type.
        // However, this does mean that we have to change the reactor index to be
        // indexed primarily by the action type, instead of the store slug.
        foreach ( $this->reactor_index[ $action_type ] as $store_slug => $hit_types ) {
            foreach ( $hooks->reaction_stores->get_children( $store_slug ) as $reaction_store ) {

                if ( ! $reaction_store instanceof WordPoints_Hook_Reaction_StoreI ) {
                    continue;
                }

                // Allowing access to stores out-of-context would lead to strange behavior.
                if ( false === $reaction_store->get_context_id() ) {
                    continue;
                }

                // TODO allow more generic queries.
                foreach ( $reaction_store->get_reactions_to_event( $event_slug ) as $reaction ) {

                    $reactor_slug = $reaction->get_reactor_slug();

                    if ( ! isset( $hit_types[ $reactor_slug ] ) ) {
                        continue;
                    }

                    $hit_type = $hit_types[ $reactor_slug ];

                    $fire = new WordPoints_Hook_Fire(
                        $action_type
                        , $event_args
                        , $reaction
                        , $hit_type
                    );

                    $this->fire_reaction( $fire );
                }
            }
        }

        foreach ( $hooks->reaction_stores->get_all() as $reaction_stores ) {
            foreach ( $reaction_stores as $reaction_store ) {

                if ( ! $reaction_store instanceof WordPoints_Hook_Reaction_StoreI ) {
                    continue;
                }

                // Allowing access to stores out-of-context would lead to strange behavior.
                if ( false === $reaction_store->get_context_id() ) {
                    continue;
                }

                // Not having this check would be silly, because we'd end up pulling
                // all of the reactions out of the database when we might not even
                // have use for any of them.
                // However, it does require that we index by store slug instead of
                // reactor slug. A side effect of this, perhaps intentional on our
                // part, is that the hit type/action type pairing for a reactor is
                // determined per reaction store then.
                $store_slug = $reaction_store->get_slug();

                if ( empty( $this->reactor_index[ $action_type ][ $store_slug ] ) ) {
                    continue;
                }

                foreach ( $reaction_store->get_reactions_to_event( $event_slug ) as $reaction ) {

                    // We moved this out of the below function because we thought it
                    // would be much better to bail earlier rather than later.
                    $reactor_slug = $reaction->get_reactor_slug();

                    if ( ! isset( $this->reactor_index[ $action_type ][ $store_slug ][ $reactor_slug ] ) ) {
                        return;
                    }

                    $hit_type = $this->reactor_index[ $action_type ][ $store_slug ][ $reactor_slug ];

                    $fire = new WordPoints_Hook_Fire(
                        $action_type
                        , $event_args
                        , $reaction
                        , $hit_type
                    );

                    $this->fire_reaction( $fire );
                }
            }
        }
    }
```

**Comment by JDGrimes** [comment]
## Reassessment

Having been through the ringer with #129, it's time to reassess this ticket. So here we go.

What are the reasons for wanting to make this externally composible in the first place? The main reason was because of the possibility of new action types and event types being introduced. If the support is hard-coded into the reactor, then those new action types would be unusable until the reactor code was updated. 

Sometimes above it seems like we're talking about new action types and event types like they'll be common things though, and they won't. That doesn't mean that the ability to map these things might not be useful for other reasons, but we should consider what those reasons are. Because at the moment, I'm actually wondering whether it is really worth it to introduce external composibility here based on the added complexity it creates. Before we roll things back though, we'd also need to consider whether all of that complexity actually goes away if this isn't externally composible, or whether this has just brought to light some things that we really should have been doing anyway (specifically I'm thinking of the logs).

**Comment by JDGrimes** [comment]
## Per reaction?

As far as going per-reaction, I think that is actually beyond the most basic goal of this ticket, and so we'd probably be better off pursuing that in its own ticket. (Assuming, of course, that we don't backtrack on this.) Here are some thoughts I've had, in case we do that:

There doesn't seem to be much use to do this per-reaction, but the fact that there might be multiple secondary response types offered by a single reactor could lead us to want to allow the reactions to choose which of those it wants to listen to. (Multiple "primary" responses really doesn't make as much sense, though guess it is possible in theory. But in that case they should probably be separate reactors, especially since that is now supported for a single reaction store. The only reason to group multiple "primary" responses together is if they wanted to share a comment secondary response, but that in itself doesn't make much sense, because if the two primary responses are so different that they have to be separate, then the secondary responses probably would be different as well.) The same thing can, of course, just be accomplished with reaction settings relating to the secondary response. However, perhaps there would be reasons we'd want to opt for the different hit types instead for the sake of the logs. Though I'm not sure what those reasons would be.

**Comment by JDGrimes** [comment]
## Other reasons for this

> Sometimes above it seems like we're talking about new action types and event types like they'll be common things though, and they won't. That doesn't mean that the ability to map these things might not be useful for other reasons, but we should consider what those reasons are.

### Switching secondary and primary responses

In the OP we said:

> > So we could match up the fire event with the reverse response, and vice versa, if we wanted to.

I'm now questioning whether that really would be possible. Or, that is, whether it would really be plausible for any implementation. It is possible, but does it "work"?

Well, let's think of it this way, using the points reactor as an example. The reverse response only reverses something that has already occurred. Firing it in absence of that thing having actually occurred will do nothing. And that's exactly what would happen if this was hooked up backward. When a post was published, no points would be removed, because they wouldn't have been awarded. _However_, after the post was depublished, points would be added, and when the post was unpublished again those points would be removed.

So at first it seems that this doesn't work, and then it does. That is, it does _seem_ that it works. In truth, it really doesn't, because what we're doing isn't actually the converse of the normal operation, which can be achieved in a much simpler manner by supplying a negative value for points and hooking things up normally. Then when a post is published points will be removed, and when the post is depublished those points will be awarded back. But in the backward-hookup scenario, this isn't what happens. The user isn't charged points for publishing a post, they're awarded points for not publishing it.

This is only thinking of the toggle events though. What about the repeatable events? Well, for them this makes no sense at all, since the reverse response really would never do anything, seeing how there would never be anything for it to reverse.

This all assumes, of course, that the secondary response of every reactor will only act relative to the primary response in some way, not independently of it. However, there is nothing about the construction of the reactors that actually demands this. On the other hand, I'm not sure that it is something that we want to encourage. But even granted that we'd ever want such a reactor, external composition is really not necessary to hook it up differently to either toggle or repeatable events. Either could be achieved using a second reactor to wrap the former one, and switch out the action types.

### Alternate responses

In a related vein, external composition would allow us to have some implementations match up alternate secondary (or even primary, I suppose) responses. However, that brings to mind again what we said [above](https://github.com/WordPoints/hooks-api/issues/109#issuecomment-218855488), arguing essentially that this doesn't make much sense, or at least isn't really necessary.

**Comment by JDGrimes** [comment]
## Added complexity

> Because at the moment, I'm actually wondering whether it is really worth it to introduce external composibility here based on the added complexity it creates. Before we roll things back though, we'd also need to consider whether all of that complexity actually goes away if this isn't externally composible, or whether this has just brought to light some things that we really should have been doing anyway (specifically I'm thinking of the logs).

The main issue is whether we'd need to log the hit type. As discussed [above](https://github.com/WordPoints/hooks-api/issues/109#issuecomment-216661754), this may not be a necessity, but if we ever wanted to query the logs by hit type for some reason, we would be unable to do so as simply or easily as if we did log the hit type. This is because we'd currently have to query based on the reactor slug and action type, or by reactor slug, action type, and reaction store slug, if we do this per implementation. There is also the issue that we might need to query multiple action types for a single hit type, and might also need to consider that the registered action type/hit type composition can change with time.

So, would we still have to worry about this if we returned to letting the reactors compose this internally? Well, we would no longer have the trouble with the action types potentially change without our notice, because our reactor would be referencing each action type internally. It would still have to query by reactor slug and action type(s), but the reactor would have all of that information at its own disposal (and I'm not sure why we'd need to run a query like that apart from the reactor or some specialized code relating to it). And of course, if we did need to worry about the hit types, we could always log them as hit meta.

In other words, this issue doesn't entirely go away, but the external composition does exacerbate it. I think we could easily get away with not logging it for now. If we find we do want to add it later, it wouldn't be too hard to do.

**Comment by JDGrimes** [comment]
Which basically means that we currently have no real reason to do this, and perhaps a reason not to, in the interest of time. So I think we'll roll this back for now, and wrap this ticket up.

**Comment by JDGrimes** [comment]
One final note here, in case we revisit this someday: the event descriptions are currently tailored for the primary action type of that type of event. So if we hooked up a reactor "backwards" in a particular implementation, the event descriptions wouldn't match up very well with what would actually take place, since the primary response of the reactor would actually be triggered by the toggle off action, not the toggle on action.
